inventory.lua

```diff
@@ -83,63 +102,109 @@ minetest.register_node("digilines:chest", {
 			return not pipeworks.connects.facingFront(i,param2)
 		end,
 		input_inventory = "main",
-		can_insert = function(pos, _, stack)
-			return can_insert(pos, stack)
+		can_insert = function(pos, _, stack, vel)
+			local ret = minetest.get_meta(pos):get_inventory():room_for_item("main", stack)
+			if not ret then
+				-- The stack cannot be accepted. It will never be passed to
+				-- insert_object, but it should be reported as a toverflow.
+				-- Here, vel = direction item is moving, which is into side.
+				local side = vector.multiply(vel, -1)
+				send_message(pos, "toverflow", stack, side)
+			end
+			return ret
 		end,
-		insert_object = function(pos, _, stack)
-			local inv = minetest.get_meta(pos):get_inventory()
-			local leftover = inv:add_item("main", stack)
-			local count = leftover:get_count()
-			if count == 0 then
-				local derpstack = stack:get_name()..' 1'
-				if not inv:room_for_item("main", derpstack) then
-					-- when you can't put a single more of whatever you just put,
-					-- you'll get a put for it, then a full
-					sendMessage(pos,"full "..maybeString(stack)..' '..tostring(count))
-				end
-			else
-				-- this happens when the chest has received two stacks in a row and
-				-- filled up exactly with the first one.
-				-- You get a put for the first stack, a put for the second
-				-- and then a overflow with the first in stack and the second in leftover
-				-- and NO full?
-				sendMessage(pos,"overflow "..maybeString(stack)..' '..tostring(count))
+		insert_object = function(pos, _, stack, vel)
+			-- Here, vel = direction item is moving, which is into side.
+			local side = vector.multiply(vel, -1)
+			local leftover = minetest.get_meta(pos):get_inventory():add_item("main", stack)
+			local leftover_count = leftover:get_count()
+			local stack_count = stack:get_count()
+			if leftover_count ~= stack_count then
+				-- Some items were added. Report them.
+				stack:set_count(stack_count - leftover_count)
+				send_message(pos, "tput", stack, side)
+				check_full(pos, "main", stack)
+			end
+			if leftover_count ~= 0 then
+				-- Some items could not be added and bounced back. Report them.
+				send_message(pos, "toverflow", leftover, side)
 			end
 			return leftover
 		end,
+		remove_items = function(pos, _, stack, dir, count)
+			-- Here, stack is the ItemStack in our own inventory that is being
+			-- pulled from, NOT the stack that is actually pulled out.
+			-- Combining it with count gives the stack that is pulled out.
+			-- Also, note that Pipeworks doesn’t pass the listname or index to
+			-- this function, so we use the ones recorded in
+			-- allow_metadata_inventory_take; because we don’t implement
+			-- tube.can_remove, Pipeworks will call
+			-- allow_metadata_inventory_take instead and will pass it the
+			-- listname and index.
+			local taken = stack:take_item(count)
+			minetest.get_meta(pos):get_inventory():set_stack(last_inventory_take_listname, last_inventory_take_index, stack)
```

Comment:
1. There is only one inventory in the chest. Why ever remember its name? You know it already!
2. There is `InvRef::remove_item` which doesn’t need index as well.

Reply:
I thought it would be more future-proof in case another list were ever added. I’m happy to get rid of that generality if you prefer, though?

Reply:
Certainly that would be better than using global (regarding the mod) variables.

Reply:
Fixed in f5387f7834da. I avoid `InvRef::remove_item` precisely because it doesn’t take `index`. Doesn’t that mean that it will always take from the first item stack that can provide the item, rather than the designated stack? That would break injector-filters in cyclic or random selection mode, which will take from different stacks of the same item at different times.